source/gridfs/gridfs-spec.md

```diff
@@ -0,0 +1,1021 @@
+  read concern).
+- **readPreference:** defaults to the read preference on the parent database (or client object if the parent database
+  has no read preference).
+- **writeConcern:** defaults to the write concern on the parent database (or client object if the parent database has no
+  write concern).
+
+The following option is transitional:
+
+- **disableMD5:** this allows users to disable MD5 when operating under FIPS restrictions. It is provided to allow a
+  transition period as drivers remove MD5 support. Until a driver removes MD5 support, drivers MUST support this option.
+  Following a driver's normal feature removal cycle, when MD5 support is removed, this option MUST be removed or
+  otherwise made into a no-op option.
+
+GridFSBucket instances are immutable. Their properties MUST NOT be changed after the instance has been created. If your
+driver provides a fluent way to provide new values for properties, these fluent methods MUST return new instances of
+GridFSBucket.
+
+### Indexes
+
+For efficient execution of various GridFS operations the following indexes MUST exist:
+
+- an index on { filename : 1, uploadDate : 1 } on the files collection
+- a unique index on { files_id : 1, n : 1 } on the chunks collection
+
+Normally we leave it up to the user to create whatever indexes they see fit, but because GridFS is likely to be looked
+at as a black box we should create these indexes automatically in a way that involves the least amount of overhead
+possible.
+
+#### Before read operations
+
+For read operations, drivers MUST assume that the proper indexes exist.
+
+#### Before write operations
+
+Immediately before the **first** write operation on an instance of a GridFSBucket class is attempted (and not earlier),
+drivers MUST:
+
+- determine if the files collection is empty using the primary read preference mode.
+- and if so, create the indexes described above if they do not already exist
+
+To determine whether the files collection is empty drivers SHOULD execute the equivalent of the following shell command:
+
+> > db.fs.files.findOne({}, { \_id : 1 })
+
+If no document is returned the files collection is empty.
+
+This method of determining whether the files collection is empty should perform better than checking the count in the
+case where the files collection is sharded.
+
+Drivers MUST check whether the indexes already exist before attempting to create them. This supports the scenario where
+an application is running with read-only authorizations.
+
+When checking whether an index exists drivers MUST compare numeric values by value even when they are of different
+types, because the actual type will depend on how the index was created (for example, indexes created using the shell
+will have double as the type for numeric values).
+
+For example, the following index specifications should all be treated as equivalent:
+
+- { filename : 1, uploadDate : 1 } // where 1 is either a 32-bit or 64-bit integer
```

Comment:
Single backticks around the document would be helpful. If you do that, you can remove the "//" before "where 1 is...".

Reply:
Done